readme.md
> Take control of your CSS keyframe animations with React

![react-jogwheel logo](https://cdn.rawgit.com/marionebl/react-jogwheel/master/react-jogwheel-banner.svg)

# react-jogwheel

`react-jogwheel` allows for declarative control of animations
by using [jogwheel](https://github.com/marionebl/jogwheel) under the hood.
It enables you to write your animations where they belong
– in your CSS - while giving you top notch control over them
at the same time.

## Installation

Fetch `react-jogwheel` via npm, install `peerDependencies`

```bash
npm install --save react-jogwheel jogwheel react react-dom
```

## Usage

```javascript
// index.jsx
import {Component} from 'react';
import {render} from 'react-dom';
import JogWheel from 'react-jogwheel';

class AnimatedElement extends Component {
  state = {
    scroll: 0,
    scrollingElement: null
  };

  handleScroll() {
    const {scrollingElement} = this.state;
    const {scrollTop} = scrollingElement;

    this.setState({
      ...this.state,
      scrollTop
    });
  }

  componentWillMount() {
    const {document} = global;
    const scrollingElement = document.scrollingElement || document.body;
    const {scrollTop} = scrollingElement;

    this.setState({
      scrollingElement,
      scrollTop
    });

    global.addEventListener('scroll', (...args) => this.handleScroll(...args));
  }

  componentWillUnmount() {
    const {document} = global;
    document.body.removeEventListener('scroll', this.handleScroll);
  }

  render() {
    <JogWheel
      {...this.props}
      progress={this.state.progress}
      />
  }
}

const element = <AnimatedElement className="animated" />;
const mount = document.querySelector('[data-mountpoint]');
render(mount, element);
```

```html
<!doctype html>
<html>
  <head>
    <title>react-jogwheel demo</title>
    <style>
      @keyframes shrink {
        from {
          transform: scale(1);
        }
        to {
          transform: scale(0.5);
        }
      }
      .animated {
        width: 100px;
        height: 100px;
        background: #000;
        border-radius: 50%;
        animation: shrink 1s both;
        animation-play-state: paused;
      }
    </style>
  </head>
  <body>
    <div data-mountpoint></div>
    <script src="index.js"></script>
  </body>
</html>
```

---

⇨ See the [examples](./source/examples/) for details.

## API

⇨ See [API Doumentation](./documentation/api.md) for technical
 interface documentation.

## Contributing

You dig `react-jogwheel` and want to submit a pull request?
Awesome! Be sure to read the [contribution guide](./contributing.md)
and you should be good to go.
Here are some notes to get you coding real quick.

```bash
git clone git@github.com:marionebl/react-jogwheel.git
cd react-jogwheel
npm install
npm start
```

---

`react-jogwheel` is built by Mario Nebl and [contributors](https://github.com/marionebl/react-jogwheel/graphs/contributors)
with :heart: and released under the [MIT License](./license.md).


changelog.md
<a name="0.1.5"></a>
## [0.1.5](https://github.com/marionebl/react-jogwheel/compare/v0.1.4...v0.1.5) (2016-05-27)




<a name="0.1.4"></a>
## [0.1.4](https://github.com/marionebl/react-jogwheel/compare/v0.1.3...v0.1.4) (2016-05-26)


### Bug Fixes

* dereference React.PropTypes for better toolings ([4f2655b](https://github.com/marionebl/react-jogwheel/commit/4f2655b))
* Fixes typescript definitions (#2) ([dcf449f](https://github.com/marionebl/react-jogwheel/commit/dcf449f)), closes [(#2](https://github.com/(/issues/2)



<a name="0.1.3"></a>
## [0.1.3](https://github.com/marionebl/react-jogwheel/compare/v0.1.2...v0.1.3) (2016-03-31)


### Bug Fixes

* **system:** publish valid npm packages ([2c45ab1](https://github.com/marionebl/react-jogwheel/commit/2c45ab1))



<a name="0.1.2"></a>
## [0.1.2](https://github.com/marionebl/react-jogwheel/compare/v0.1.1...v0.1.2) (2016-03-12)


### Bug Fixes

* move usage of react node reference to componentDidMount ([b079cdd](https://github.com/marionebl/react-jogwheel/commit/b079cdd)), closes [#1](https://github.com/marionebl/react-jogwheel/issues/1)



<a name="0.1.1"></a>
## [0.1.1](https://github.com/marionebl/react-jogwheel/compare/v0.1.0...v0.1.1) (2016-03-01)


### Bug Fixes

* support react elements in props.component ([c85c592](https://github.com/marionebl/react-jogwheel/commit/c85c592))



<a name="0.1.0"></a>
# 0.1.0 (2016-02-26)
* Initial reelease

---

patternplate-core is built by Mario Nebl and [contributors](./documentation/contributors.md)
with :heart: and released under the [MIT License](./license.md).


source/index.js
import debuglog from 'debuglog';
import * as React from 'react';
import {Component} from 'react';

import init from './library/init-jogwheel';
import onSample from './library/on-sample';

const log = debuglog('react-jogwheel');

/**
 * noop function
 * @returns {null} null
 * @access private
 */
const noop = () => {};

export default class ReactJogWheel extends Component {
	/**
	 * React wrapper component for JogWheel
	 * @author Mario Nebl
	 * @copyright 2016 Mario Nebl
	 * @license MIT
	 * @version 0.1.1
	 * @fileoverview Take control of your CSS keyframe animations with react
	 * @see https://github.com/marionebl/jogwheel
	 * @typedef {Object} ReactJogwheel
	 * @property {element|string|function} component='div' to render
	 * @example
	 * // Render with alternative tagName
	 * <JogWheel component="span" />
	 * // Render other React component
	 * <JogWheel component={Toolbar} />
	 *
	 * @property {node|Array<node>} children to render
	 * @example
	 * // Render with children
	 * <JogWheel>
	 *  <span className="child" />
	 * </JogWheel>
	 *
	 * @property {number} progress of the applied animation
	 * @example
	 * // Render with progress set to mid-animation
	 * <JogWheel progress={0.5} />
	 *
	 * @property {function} onProgress=() => {} callback fired progress changes
	 * @example
	 * // Render with onProgress callback
	 * onProgress (...args) => {
	 *   console.log(...args);
	 * }
	 * <JogWheel {...{onProgress}} />
	 *
	 * @property {number} duration of the animation in milliseconds, overrides `animation-duration`
	 * @example
	 * // Force animation-duration to 100ms
	 * <JogWheel duration={100} />
	 *
	 * @property {number} delay of the animation in milliseconds, overrides `animation-delay`
	 * @example
	 * // Force animation-delay to 1000ms
	 * <JogWheel delay={1000} />
	 *
	 * @property {number} iterationCount of the animation, overrides `animation-iteration-count`
	 * @example
	 * // Force animation-iteration-count to 42
	 * <JogWheel iterationCount={42} />
	 *
	 * @property {string} timingFunction of the animation overrides `animation-timing-function`
	 * @example
	 * // Force animation-timing-function to 'ease-in-out'
	 * <JogWheel timingFunction="ease-in-out" />
	 *
	 * @property {boolean} playing state of the animation
	 * @example
	 * // Force animation-play-state to 'paused'
	 * <JogWheel playing={false} />
	 * // Force animation-play-state to 'playing'
	 * <JogWheel playing />
	 *
	 */
	static propTypes = {
		children: React.PropTypes.oneOfType([
			React.PropTypes.arrayOf(React.PropTypes.node),
			React.PropTypes.node
		]),
		component: React.PropTypes.oneOfType([
			React.PropTypes.element,
			React.PropTypes.func,
			React.PropTypes.string
		]),
		delay: React.PropTypes.number,
		duration: React.PropTypes.number,
		iterationCount: React.PropTypes.number,
		onProgress: React.PropTypes.func,
		playing: React.PropTypes.bool,
		progress: React.PropTypes.number,
		timingFunction: React.PropTypes.string
	};

	/**
	 * Internal defaultProps
	 * * render as `div` per default
	 * * onProgress is a `noop` func per default
	 *
	 * @type {Object}
	 * @access private
	 */
	static defaultProps = {
		component: 'div',
		onProgress: noop
	};

	/**
	 * Internal JogWheel instance attached after `render`
	 *
	 * @type {object}
	 * @access private
	 */
	instance = null;

	/**
	 * Internal node reference
	 * @type {HTMLElement}
	 * @access private
	 */
	node = null;

	/**
	 * Internal component state holding JogWheel styles for
	 * polyfilled integration cases
	 *
	 * @type {Object}
	 * @access private
	 */
	state = {
		style: {}
	};

	constructor(...args) {
		super(...args);
		this.ref = this.ref.bind(this);
	}

	/**
	 * Save the node reference to component instance
	 * @param  {HTMLElement} node refernce to save
	 * @return {void}
	 * @access private
	 */
	ref(node) {
		log('::ref', node);
		this.node = node;
	}

	/**
	 * Detach the sampler loop when unmounting
	 * @returns {null} null
	 * @access private
	 */
	componentWillUnmount() {
		log('::componentWillUnmount');
		global.cancelAnimationFrame(this.sampler);
	}

	/**
	 * Set up after mount
	 * * seek on attached JogWheel instance
	 * * play/pause on attached JogWheel instance
	 * * attach sampler loop
	 * @returns {null} null
	 * @access private
	 */
	componentDidMount() {
		log('::componentDidMount');
		this.instance = init(this);

		if (typeof this.props.progress === 'number') {
			this.instance.seek(this.props.progress);
		}

		if (this.props.playing) {
			this.instance.play();
		} else {
			this.instance.pause();
		}

		/**
		 * Internal Sampler instance attached after `componentDidMount`
		 * @type {object}
		 * @returns {null} null
		 * @access private
		 */
		this.sampler = onSample(this);
	}

	/**
	 * Apply changed props to JogWheel instance
	 * @param {object} props properties about to receive
	 * @returns {null} null
	 * @access private
	 */
	componentWillReceiveProps(props) {
		log('::componentWillReceiveProps', props);
		if (typeof props.progress === 'number' && props.progress !== this.props.progress) {
			this.instance.seek(props.progress);
		}

		if (props.playing !== this.props.playing) {
			if (props.playing) {
				this.instance.play();
			} else {
				this.instance.pause();
			}
		}
	}

	/*
	 * Render the component
	 * @returns {null} null
	 * @access private
	 */
	render() {
		log('::render', this.props, this.state);
		const Component = this.props.component;
		return (
			<Component
				{...this.props}
				ref={this.ref}
				style={this.state.style}
				>
				{this.props.children}
				</Component>
		);
	}
}


source/library/init-jogwheel.js
import {Component} from 'react';
import {findDOMNode} from 'react-dom';
import JogWheel from 'jogwheel';

/**
 * Attach a JogWheel instance to component
 * @param  {object} component React component to attach to
 * @return {object} jogwheel instance
 * @access private
 */
export default component => {
	const {node} = component;
	const element = node instanceof Component ?
		findDOMNode(node) :
		node;

	return JogWheel.create(element, {
		/**
		 * Render the new frame on component
		 * @param  {*} _             placeholder
		 * @param  {string} propertyName  name of the property
		 * @param  {*} propertyValue value of the property
		 * @return {null}                 null
		 * @access private
		 */
		render(_, propertyName, propertyValue) {
			component.setState({
				style: {
					...component.state.style,
					[propertyName]: propertyValue
				}
			});
			return propertyValue;
		}
	});
};


source/examples/scroll.js
import 'web-animations-js';

import * as React from 'react';
import {Component, PropTypes} from 'react';
import * as ReactDOM from 'react-dom';

import JogWheel from '../';

class Toolbar extends Component {
	render() {
		return (
			<div className="toolbar">
				Toolbar
			</div>
		);
	}
}

class AnimatedToolbar extends Component {
	static propTypes = {
		progress: React.PropTypes.number
	};

	render() {
		return (
			<JogWheel
				component={Toolbar}
				progress={this.props.progress}
				/>
		);
	}
}

class AnimatedScrollContainer extends Component {
	static propTypes = {
		children: PropTypes.node
	};

	render() {
		return (
			<JogWheel
				component="div"
				{...this.props}
				className="scroll-container"
				>
				{this.props.children}
			</JogWheel>
		);
	}
}

class Application extends Component {
	state = {
		scroll: 0,
		scrollElement: null
	};

	handleScroll() {
		this.setState({
			...this.state,
			scroll: this.state.scrollElement.scrollTop
		});
	}

	componentWillMount() {
		const {document} = global;
		const scrollElement = document.scrollingElement || document.body;

		this.setState({
			scrollElement,
			scroll: scrollElement.scrollTop
		});

		global.addEventListener('scroll', (...args) => this.handleScroll(...args));
	}

	componentWillUnmount() {
		const {document} = global;
		document.body.removeEventListener('scroll', this.handleScroll);
	}

	render() {
		return (
			<div className="application">
				<AnimatedToolbar progress={this.state.scroll / 1000} />
				<AnimatedScrollContainer progress={this.state.scroll / 1000}>
					<div className="demo-content">
						Demo
					</div>
					<div className="demo-content">
						Demo
					</div>
					<div className="demo-content">
						Demo
					</div>
					<div className="demo-content">
						Demo
					</div>
					<div className="demo-content">
						Demo
					</div>
					<div className="demo-content">
						Demo
					</div>
					<div className="demo-content">
						Demo
					</div>
				</AnimatedScrollContainer>
			</div>
		);
	}
}

/**
 * Execute the example
 * @param {object} [window=global] Global window context
 * @param {object} [document=global.document] Global document context
 * @returns {null} null
 */
function main(window = global, document = global.document) {
	// create mountpoint
	const mount = document.createElement('div');
	document.body.appendChild(mount);

	const style = document.createElement('style');
	style.innerText = `
		* {
			box-sizing: border-box;
		}

		html,
		body {
			font-family: Arial, sans-serif;
			color: #fff;
			margin: 0;
			padding: 0;
		}

		@keyframes shrink {
			from {
				height: 200px;
			}
			to {
				height: 100px;
			}
		}
		.toolbar {
			position: fixed;
			top: 0;
			left: 0;
			display: flex;
			align-items: center;
			justify-content: center;
			width: 100%;
			height: 200px;
			color: #fff;
			background: #000;
			animation: shrink 1s both;
			animation-play-state: paused;
		}

		.demo-content {
			display: flex;
			align-items: center;
			justify-content: center;
			width: 100%;
			height: 50vh;
			margin-bottom: 5vh;
			background: #ddd;
			color: #333;
		}
	`;

	document.body.appendChild(style);

	// Render stuffs
	ReactDOM.render(<Application />, mount);
}

main();
